tools/autograd/gen_variable_type.py

```diff
@@ -494,39 +500,40 @@ def format_trace_op_name(declaration):
 def format_trace_inputs(declaration):
     def dispatch_trace_input(arg_spec):
         name, value, simple_type, nullable = arg_spec
-        if declaration['use_c10_dispatcher'] == 'full':
-            if value == "options":
-                value = "TensorOptions().dtype(dtype).layout(layout).device(device).pinned_memory(pin_memory)"
-        else:
-            assert declaration['use_c10_dispatcher'] == 'with_codegenerated_unboxing_wrapper'
         # XXX: For arg that have type of Tensor?[], tracer will pass allow_undefined to addInputs
         if simple_type == 'TensorList' and nullable:
             return '''jit::tracer::addInputs(node, "{}", {}, {});'''.format(name, value, "true")
         else:
-            return ADD_TRACE_INPUT.substitute(name=name, input=value)
+            if value == "options":
+                result = ""
+                result += ADD_TRACE_INPUT.substitute(name="dtype", input="optTypeMetaToScalarType(options.dtype_opt())") + "\n"
+                result += ADD_TRACE_INPUT.substitute(name="layout", input="options.layout_opt()") + "\n"
+                result += ADD_TRACE_INPUT.substitute(name="device", input="options.device_opt()") + "\n"
+                result += ADD_TRACE_INPUT.substitute(name="pin_memory", input="options.pinned_memory_opt()")
+                return result
+            else:
+                return ADD_TRACE_INPUT.substitute(name=name, input=value)
 
-    trace_inputs = declaration['arguments']
+    if declaration['use_c10_dispatcher'] == 'full':
+        trace_inputs = declaration['schema_order_arguments']
+        trace_input_spec = [(i['name'], i['name'], i['type'], i.get('is_nullable')) for i in trace_inputs]
+    else:
+        trace_inputs = declaration['arguments']
+        trace_input_spec = [(i['name'], i['name'], i['simple_type'], i.get('is_nullable')) for i in trace_inputs]
```

Comment:
I don't know much about the difference between `simple_type` and `type`. `simple_type` seems to be only present for non-c10-full ops, but for them it has to be used instead of `type`, otherwise it doesn't compile.